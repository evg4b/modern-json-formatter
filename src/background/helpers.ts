export const wait = <T>(request: IDBRequest<T>): Promise<T> => new Promise((resolve, reject) => {
  request.onsuccess = () => resolve(request.result);
  request.onerror = () => reject(request.error);
});
