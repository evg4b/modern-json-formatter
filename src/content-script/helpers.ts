import { ErrorNode } from '@worker-core';

export const throws = (value?: string): never => {
  throw new Error(value ?? 'Unexpected value');
};

export const query = <T extends HTMLElement = HTMLElement, D extends HTMLElement = HTMLElement>(
  value: D,
  selector: string,
): T => value.querySelector<T>(selector) ?? throws(`Element ${ selector } not found`);

const types: string[] = ['raw', 'query', 'formatted'] satisfies TabType[];

export function assetTabType(s?: string | null): asserts s is TabType {
  if (s && !types.includes(s)) {
    throw new Error(`Invalid tab type '${ s }'`);
  }
}

export const isNotNil = <T>(value: T | null | undefined): value is T => {
  return value !== null && value !== undefined;
};

export const isErrorNode = (node: unknown): node is ErrorNode => {
  return !!node && typeof node === 'object' && 'type' in node && node.type === 'error';
}
