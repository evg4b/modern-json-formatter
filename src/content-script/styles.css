@import "./styles/host.css";
@import "./styles/code.css";
