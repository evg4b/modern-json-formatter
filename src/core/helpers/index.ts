export * from './trim';
export * from './is-in-viewport';
