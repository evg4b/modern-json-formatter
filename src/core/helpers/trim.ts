export const trim = (value: string, chars: string = ' \t\n\r') => {
  const pattern = new RegExp(`^[${chars}]+|[${chars}]+$`, 'g');
  return value.replace(pattern, '');
};
