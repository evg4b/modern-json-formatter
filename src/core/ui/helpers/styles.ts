import { createElement } from '@core/dom';

export const registerStyle = (shadowRoot: ShadowRoot, styles: string) => {
  const styleNode = document.createElement('style');
  styleNode.textContent = styles;
  styleNode.setAttribute('type', 'text/css');
  styleNode.setAttribute('rel', 'stylesheet');
  styleNode.setAttribute('role', 'presentation');
  shadowRoot.appendChild(styleNode);

  return styleNode;
};

export const registerStyleLink = (shadowRoot: ShadowRoot, url: string) => {
  const linkNode = createElement({
    element: 'link',
    attributes: {
      href: url,
      type: 'text/css',
      rel: 'stylesheet',
      role: 'presentation',
    },
  });

  shadowRoot.appendChild(linkNode);

  return linkNode;
};
