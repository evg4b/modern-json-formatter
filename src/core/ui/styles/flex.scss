.d-flx {
  display: flex;
}

.flx-row {
  flex-direction: row;
}

.flx-column {
  flex-direction: column;
}

.flx-center {
  align-items: center;
}
