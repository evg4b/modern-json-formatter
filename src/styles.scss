$indentation: 30px;

@font-face {
  font-family: 'Monaco';
  font-style: normal;
  font-weight: normal;
  src: local('Monaco'), url('chrome-extension://__MSG_@@extension_id__/Monaco.woff') format('woff');
}

:host {
  display: flex;
  flex-direction: column;
  padding: $indentation;
  font-family: "Monaco", Arial, sans-serif;
  background: #282828;
  color: #eeeeee;
  font-size: 14px;
  width: 100%;
  box-sizing: border-box;

  pre {
    margin: 0;
    padding: 0;
    text-wrap: pretty;
    word-wrap: break-word;
    font-family: "Monaco", Arial, sans-serif;
    font-size: 14px;
  }

  & > .root-container.formatted {
    & > .raw-json-container {
      display: none;
    }

    & > .formatted-json-container {
      display: block;
    }

    & > .query-json-container {
      display: none;
    }
  }

  & > .root-container.raw {
    & > .raw-json-container {
      display: block;
    }

    & > .formatted-json-container {
      display: none;
    }

    & > .query-json-container {
      display: none;
    }
  }

  & > .root-container.query {
    & > .raw-json-container {
      display: none;
    }

    & > .formatted-json-container {
      display: none;
    }

    & > .query-json-container {
      display: block;
    }
  }

  & > .root-container > {
    .raw-json-container {
      display: none;
    }

    .formatted-json-container {
      display: block;
    }
  }

  .toolbox-container {
    display: flex;
    justify-content: space-between;
    margin-bottom: 10px;
    position: fixed;
    top: 15px;
    right: 30px;

    input {
      min-width: 200px;
      width: 30vw;
    }
  }
}

.button-container {
  display: flex;
  flex-direction: row;
}

pre {
  margin: 0;
  padding: 0;
}

.bracket {
  color: #eeeeee;
}

.object {
  display: inline;

  .inner {
    padding-left: $indentation;
    display: block;
    border-left: 1px dotted #3b3b3b;

    .property {
      display: block;
      margin: 2px 0 2px 0;

      &.collapsed {
        & > .object,
        & > .array {
          .inner {
            display: none;
          }

          .ellipsis {
            display: inline;
          }

          & > .items-count,
          & > .properties-count {
            display: inline;
          }
        }
      }
    }
  }
}

.root {
  display: block;

  &.padding {
    margin-left: 17px;
  }

  &.collapsed {
    & > .object,
    & > .array {
      .inner {
        display: none;
      }

      .ellipsis {
        display: inline;
      }

      & > .items-count,
      & > .properties-count {
        display: inline;
      }
    }
  }
}

.array {
  display: inline;

  .inner {
    padding: 2px 0 2px $indentation;
    display: block;

    .item {
      display: block;
      margin: 2px 0 2px 0;

      &.collapsed {
        & > .object,
        & > .array {
          .inner {
            display: none;
          }

          .ellipsis {
            display: inline;
          }

          & > .items-count,
          & > .properties-count {
            display: inline;
          }
        }
      }
    }
  }
}

.items-count,
.properties-count {
  color: #717171;
  padding-left: 15px;
  display: none;
  user-select: none;
}

.string {
  color: #5dd9ff;
  word-wrap: break-word;
}

.boolean,
.number {
  color: #9f84ff;
}

.key {
  color: #7dadf9;
  word-wrap: break-word;
}

.null {
  color: #717171;
}

.colon {
  color: #eeeeee;
  padding: 0 7px 0 0;
}

.toolbox-container {
  gap: 10px;

  button {
    &:first-child {
      border-bottom-left-radius: 5px;
      border-top-left-radius: 5px;
    }

    &:last-child {
      border-bottom-right-radius: 5px;
      border-top-right-radius: 5px;
    }

    &:hover {
      background: #525252;
    }

    &.active {
      background: #2c2c2c;
      color: #b4b2b2;
      cursor: default;
    }

    padding: 5px 10px;
    background: #3b3b3b;
    color: #eeeeee;
    border: none;
    cursor: pointer;
    font-size: 14px;
    font-weight: 500;
    transition: background 0.3s;
    min-width: 41px;
    display: block;
    user-select: none;
  }
}

.toggle {
  display: inline-block;
  width: 17px;
  height: 17px;
  cursor: pointer;
  margin-left: -17px;

  &:before {
    content: '';
    display: inline-block;
    border-style: solid;
    border-width: 5px 0 5px 8px;
    border-color: transparent transparent transparent currentColor;
    transform: translateX(4px) rotate(90deg);
    transition: transform 0.1s;
    color: #575757;
  }

  &:hover:before {
    color: #a6a6a6;
  }
}

.ellipsis {
  &:before {
    content: '...';
    color: #717171;
    padding: 0 5px;
  }

  display: none;
  user-select: none;
}

.object,
.array {
  .bracket + .bracket {
    margin-left: 5px;
  }
}

.collapsed {
  & > .toggle:before {
    transform: translateX(4px) rotate(0deg);
  }

  & > .items-count,
  & > .properties-count {
    display: inline;
  }
}

.error {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 15px;
  justify-content: space-around;
  color: #6e6e6e;
  user-select: none;
  pointer-events: none;
  padding: 5em 0;

  img {
    width: 80%;
    max-width: 250px;
    user-select: none;
    pointer-events: none;
  }

  .message {
    display: flex;
    flex-direction: column;
    align-items: center;
    pointer-events: none;
  }
}
