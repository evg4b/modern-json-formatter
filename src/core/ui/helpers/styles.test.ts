import { registerStyle, registerStyleLink } from '@core/ui/helpers/styles';
import { beforeAll, describe, expect, it } from '@jest/globals';

describe('registerStyle', () => {
  let shadowRoot: ShadowRoot;
  let styleElement: HTMLStyleElement;

  beforeAll(() => {
    const host = document.createElement('div');
    shadowRoot = host.attachShadow({ mode: 'open' });
    styleElement = registerStyle(shadowRoot, `
      :host {
        display: flex;
        flex-direction: column;
        gap: 1rem;
        align-items: center;
        justify-content: center;  
      }
    `);
  });

  it('should return the style node', () => {
    expect(styleElement).toMatchSnapshot();
  });

  it('should attach the style node to the shadow root', () => {
    expect(shadowRoot.innerHTML).toMatchSnapshot();
  });
});

describe('registerStyleLink', () => {
  let shadowRoot: ShadowRoot;
  let linkElement: HTMLLinkElement;

  beforeAll(() => {
    const host = document.createElement('div');
    shadowRoot = host.attachShadow({ mode: 'open' });
    linkElement = registerStyleLink(shadowRoot, `/css/content-styles.css`);
  });

  it('should return the link node', () => {
    expect(linkElement).toMatchSnapshot();
  });

  it('should render the lint to style', () => {
    expect(shadowRoot.innerHTML).toMatchSnapshot();
  });
});
