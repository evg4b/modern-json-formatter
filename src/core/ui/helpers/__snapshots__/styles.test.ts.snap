// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`registerStyle should attach the style node to the shadow root 1`] = `
"<style type="text/css" rel="stylesheet" role="presentation">
      :host {
        display: flex;
        flex-direction: column;
        gap: 1rem;
        align-items: center;
        justify-content: center;  
      }
    </style>"
`;

exports[`registerStyle should return the style node 1`] = `
<style
  rel="stylesheet"
  role="presentation"
  type="text/css"
>
  
      :host {
        display: flex;
        flex-direction: column;
        gap: 1rem;
        align-items: center;
        justify-content: center;  
      }
    
</style>
`;

exports[`registerStyleLink should render the lint to style 1`] = `"<link href="/css/content-styles.css" type="text/css" rel="stylesheet" role="presentation">"`;

exports[`registerStyleLink should return the link node 1`] = `
<link
  href="/css/content-styles.css"
  rel="stylesheet"
  role="presentation"
  type="text/css"
/>
`;
